custom/enikshay/tasks.py

```diff
@@ -0,0 +1,180 @@
+    """
+    index by day of adherence_date datetime
+    """
+    by_date = defaultdict(list)
+    for case in adherence_cases:
+        adherence_date = parse_datetime(case.dynamic_case_properties().get('adherence_date')).date()
+        by_date[adherence_date].append(case)
+    return by_date
+
+
+class EpisodeUpdate(object):
+
+    def __init__(self, episode_case, case_updater):
+        self.episode = episode_case
+        self.case_updater = case_updater
+
+    def get_property(self, property):
+        return self.episode.dynamic_case_properties().get(property)
+
+    def get_latest_adherence_case_for_episode(self):
+        """
+        return open case of type 'adherence' reversed-indexed to episode and
+            with latest 'adherence_date' property
+        """
+        case_accessor = CaseAccessors(self.case_updater.domain)
+        indexed_cases = case_accessor.get_reverse_indexed_cases([self.episode.case_id])
+        # sometime far back
+        latest_date = pytz.UTC.localize(datetime.datetime(1990, 1, 1))
+        latest_case = None
+        for case in indexed_cases:
+            adherence_date = parse_datetime(case.dynamic_case_properties().get('adherence_date'))
+            if (not case.closed and
+               case.type == CASE_TYPE_ADHERENCE and
+               adherence_date > latest_date):
+                latest_date = adherence_date
+                latest_case = case
+        return latest_case
+
+    def update_json(self):
+        adherence_schedule_date_start = parse_datetime(self.get_property('adherence_schedule_date_start'))
+        if not adherence_schedule_date_start:
+            # adherence schedule hasn't been selected, so no update necessary
+            return {}
+
+        if adherence_schedule_date_start > self.case_updater.purge_date:
+            return {
+                'aggregated_score_date_calculated': adherence_schedule_date_start - datetime.timedelta(days=1),
+                'expected_doses_taken': 0,
+                'aggregated_score_count_taken': 0
+            }
+        else:
+            update = {}
+            adherence_case = self.get_latest_adherence_case_for_episode()
+            if adherence_case:
+                adherence_date = parse_datetime(adherence_case.dynamic_case_properties().get('adherence_date'))
+            else:
+                adherence_date = None
+            if not adherence_case:
+                update["aggregated_score_date_calculated"] = self.case_updater.purge_date
+                update["aggregated_score_count_taken"] = 0
+            elif adherence_date < self.case_updater.purge_date:
+                update["aggregated_score_date_calculated"] = adherence_date
+            else:
+                update["aggregated_score_date_calculated"] = self.case_updater.purge_date
+
+            # calculate 'aggregated_score_count_taken'
+            if adherence_case:
+                adherence_cases = get_adherence_cases_between_dates_from_episode(
```

Comment:
if we are going to load all the adherence cases from the episode to get the latest one, we should at least reuse that list here to not have to go back to the DB. Again, moving to elasticsearch would probably help a lot for performance of this.